snapshotter/devmapper/devmapper.go

```diff
@@ -0,0 +1,319 @@
+	if err != nil {
+		return nil, err
+	}
+
+	snap, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		return nil, complete(ctx, trans, err)
+	}
+
+	if len(snap.ParentIDs) == 0 {
+		deviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating new thin device '%s'", deviceName)
+
+		deviceID, err := dm.pool.CreateThinDevice(deviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create thin device for snapshot %s", snap.ID)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created thin device with id %d", deviceID)
+		if err := dm.mkfs(ctx, deviceName); err != nil {
+			return nil, complete(ctx, trans, err)
+		}
+	} else {
+		parentDeviceName := dm.getDeviceName(snap.ParentIDs[0])
+		snapDeviceName := dm.getDeviceName(snap.ID)
+		log.G(ctx).Debugf("creating snapshot device '%s' from '%s'", snapDeviceName, parentDeviceName)
+
+		snapDeviceID, err := dm.pool.CreateSnapshotDevice(parentDeviceName, snapDeviceName, dm.config.BaseImageSizeBytes)
+		if err != nil {
+			log.G(ctx).WithError(err).Errorf("failed to create snapshot device from parent %s", parentDeviceName)
+			return nil, complete(ctx, trans, err)
+		}
+
+		log.G(ctx).Debugf("created snapshot device with id %d", snapDeviceID)
+	}
+
+	mounts := dm.buildMounts(snap)
+
+	// This is needed only to pass containerd's snapshotter suite tests
+	_ = mount.WithTempMount(ctx, mounts, func(root string) error {
+		return os.Remove(filepath.Join(root, "lost+found"))
+	})
+
+	return mounts, complete(ctx, trans, nil)
+}
+
+func (dm *Snapshotter) mkfs(ctx context.Context, deviceName string) error {
+	args := []string{
+		"-E",
+		"nodiscard,lazy_itable_init=0,lazy_journal_init=0",
+		dm.pool.GetDevicePath(deviceName),
+	}
+
+	log.G(ctx).Debugf("mkfs.ext4 %s", strings.Join(args, " "))
+	output, err := exec.Command("mkfs.ext4", args...).CombinedOutput()
+	if err != nil {
+		log.G(ctx).WithError(err).Errorf(
+			"failed to write fs: %s\ndmesg: %s\n",
+			string(output),
+			string(dmesg.Dmesg(256)))
+		return err
+	}
+
+	log.G(ctx).Debugf("mkfs:\n%s", string(output))
+	return nil
+}
+
+func (dm *Snapshotter) getDeviceName(snapID string) string {
+	// Add pool name as prefix to avoid collisions with devices from other pools
+	return fmt.Sprintf("%s-snap-%s", dm.config.PoolName, snapID)
+}
+
+func (dm *Snapshotter) getDevicePath(snap storage.Snapshot) string {
+	name := dm.getDeviceName(snap.ID)
+	return dm.pool.GetDevicePath(name)
+}
+
+func (dm *Snapshotter) buildMounts(snap storage.Snapshot) []mount.Mount {
+	var options []string
+
+	if snap.Kind != snapshots.KindActive {
+		options = append(options, "ro")
+	}
+
+	mounts := []mount.Mount{
+		{
+			Source:  dm.getDevicePath(snap),
+			Type:    fsTypeExt4,
+			Options: options,
+		},
+	}
+
+	return mounts
+}
+
+func complete(ctx context.Context, trans storage.Transactor, err error) error {
+	if err != nil {
```

Comment:
Generally with if else statements it is more readable to return early in the if statement, if possible. Since this is more of a stylistic concern, what does `containerd` usually do? 

Something like this
```go
if err != nil {
    if terr := trans.Rollback(); terr != nil {
         log.G(ctx).WithError(terr).Error("failed to rollback transaction")
    }
    return err
}

if terr := trans.Commit(); terr != nil {
    log.G(ctx).WithError(terr).Error("failed to commit transaction")
}
return nil
```

Reply:
Generally I would agree that we have to have as less branches as possible, but in this case there are two equally important / symmetric blocks of code, and they look more clear that we do either `Commit` or `Rollback`

Reply:
That's fair. It's more stylistic than anything else.